Blazor Open Links in Browser with Configurability

TanayParikh opened the pull request:
- Opens link in external browser by default instead of within the WebView
- Optionally allows opening links in the Blazor WebView by attaching to the `ExternalNavigationStarting` event and setting the `ExternalLinkNavigationPolicy.OpenInWebView`.
  - Likewise allows cancelling navigation via `ExternalLinkNavigationPolicy.CancelNavigation`
- Always opens `_blank` target links in external browser

Status:
- Maui
  - [x] Windows
  - [x] Android
  - [x] iOS
  - [x] Mac Catalyst
- [x] WinForms
- [x] WPF

Fixes: https://github.com/dotnet/maui/issues/4357
Fixes: https://github.com/dotnet/maui/issues/4338

TanayParikh committed:
Blazor Android Open Links in Browser with Configurability

TanayParikh commented on src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:
```diff
@@ -20,25 +22,37 @@ protected WebKitWebViewClient(IntPtr javaReference, JniHandleOwnership transfer)
 				var uri = new Uri(requestUri);
-				if (uri.Host == "0.0.0.0")
+
+				if (uri.Host == "0.0.0.0" &&
+					view is not null && 
+					request is not null && 
+					request.IsRedirect && 
+					request.IsForMainFrame)
 				{
 					view.LoadUrl(uri.ToString());
 					return true;
 				}
```
It's not totally clear to me when/why this code path gets exercised. Could anyone offer insights?

Eilon replied:
Maybe setting `window.location` in JS? Just a guess. Much of this code is likely copied from the non-Blazor WebView, so it could be for scenarios we don't really have in Blazor.

Eilon commented on src/BlazorWebView/src/Maui/Android/WebKitWebViewClient.cs:
```diff
@@ -20,25 +22,37 @@ protected WebKitWebViewClient(IntPtr javaReference, JniHandleOwnership transfer)
 				var uri = new Uri(requestUri);
-				if (uri.Host == "0.0.0.0")
+
+				if (uri.Host == "0.0.0.0" &&
```
I realize this code was already here, but maybe we can store the string `0.0.0.0` in some field and use it everywhere? Example: `private const string AppHostAddress = "0.0.0.0";`

TanayParikh committed:
Blazor Windows Open Links in Browser with Configurability (#4680)

* Blazor Windows Open Links in Browser with Configurability

Windows portion of #4338

* TryCreate URI

TanayParikh committed:
PR Feedback

(cherry picked from commit 35f637ee0371044f1df1f6ea5092ef69e8a14215)

TanayParikh commented on Microsoft.Maui.sln:
```diff
@@ -200,6 +200,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "MauiBlazorWebView.DeviceTes
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "SharedSource", "SharedSource", "{4F2926C8-43AB-4328-A735-D9EAD699F81D}"
 	ProjectSection(SolutionItems) = preProject
+		src\BlazorWebView\src\SharedSource\ExternalLinkMode.cs = src\BlazorWebView\src\SharedSource\ExternalLinkMode.cs
```
Brought this out to shared source so it could be used for future WPF/Winforms implementations.

TanayParikh commented:
> I was expecting our API to be something more like a virtual method you can override to receive the URL that is being navigated to, and return a flag to say whether (for this specific URL) it should be allowed to open internally.

I think this makes sense. @Eilon  had also suggested an event based option for us to consider but I was running into some issues with how the different platforms handle navigation at different stages. 

I'm leaning towards an implementation which just takes an optional delegate as the arg (instead of ExternalLinkMode). If the delegate is provided then we simply call it when deciding whether or not to open the link within the WebView. How does this sound?

Eilon commented:
How does the app developer pass in the delegate? Presumably it's an event on the control?

TanayParikh commented:
> How does the app developer pass in the delegate?

During Blazor WebView initialization:

https://github.com/dotnet/maui/blob/d3f08ea078101b815c02e3ee257d29e0c7430873/src/Controls/samples/Controls.Sample/Pages/Blazor/BlazorPage.xaml.cs#L32-L39

is what I had in mind.

TanayParikh committed:
Merge branch 'main' into taparik/androidAnchorTagConfigurability

TanayParikh commented:
> @TanayParikh - sure, but using what syntax? An event? Events are the canonical pattern for controls.

The issue with doing an `EventCallback<ExternalLinkEventArgs>` is the `InvokeAsync` to actually call the callback. This doesn't work with the synchronous event handlers we have for actually working with navigation on the individual platforms (we're unable to await in those contexts).

I was proposing just using a non-event delegate function, granted, it wouldn't be the canonical approach.

SteveSandersonMS commented:
> I was proposing just using a non-event delegate function, granted, it wouldn't be the canonical approach.

Sounds reasonable to me. I agree that `EventCallback<T>` would be sketch as it implies the option to make choices asynchronously.

We could do it via a plain C# event in which we pass some new `NavigationInfo` parameter giving the new URL and exposing an `OpenInWebView` callback, e.g.:

```cs
myBlazorWebView.OnExternalNavigationStarting += (sender, eventArgs) =>
{
    if (ThisLooksLikeATrustworthyUrl(eventArgs.Url))
    {
        eventArgs.OpenInWebView();
    }
};
```

... but TBH that seems so much weirder than something like:

```cs
// If not set, default to NavigationPolicy.OpenInSystemBrowser
myBlazorWebView.NavigationPolicy = navigationInfo =>
{
    return ThisLooksLikeATrustworthyUrl(navigationInfo.Url) ? NavigationPolicy.OpenInWebView :  NavigationPolicy.OpenInSystemBrowser;
};
```

We should check whether the new event/callback fires for *all* navigations (including ones handled by Blazor's client-side router), or only ones that aren't handled by the client-side router. The API name should account for this, e.g., call it `ExternalNavigationStarting` or just `NavigationStarting` depending on this distinction.

TanayParikh committed:
OnExternalNavigationStarting

TanayParikh commented:
Thanks @SteveSandersonMS, I've updated via https://github.com/dotnet/maui/pull/4645/commits/3c7c2d0a54858b76ef2a1daab6d5a877a2f0b9dc.

> We should check whether the new event/callback fires for all navigations (including ones handled by Blazor's client-side router), or only ones that aren't handled by the client-side router.

We're specifically looking at "external" links (defined by any `Uri.Host` other than `0.0.0.0`.

Ex. https://github.com/dotnet/maui/pull/4645/files#diff-491a4b35962629acb0b5b087bac0135dc974d6f395288b07eae1f8e0de608a00R47

TanayParikh committed:
Pranav Points

Eilon commented:
> ... but TBH that seems less ergonomic than your regular-delegate proposal. Maybe something like:

That might be true in "regular code" but when working with controls the only "ergonomic" pattern is to use events. It also won't work in any tooling (e.g. in a property grid, XAML Intellisense, etc.).

TanayParikh committed:
Event based approach

TanayParikh committed:
Info -> EventArgs

TanayParikh commented:
> when working with controls the only "ergonomic" pattern is to use events. It also won't work in any tooling (e.g. in a property grid, XAML Intellisense, etc.).

Thanks for the clarification, I believe https://github.com/dotnet/maui/pull/4645/commits/bdfabfd3cf4b7238bb597756d17e3e6705f0c8c0 should resolve this. Had to do a sort of awkward workaround using an `Action<ExternalLinkNavigationInfo>` to get around event invocation from the source, but it gets the job done and it works cross platform (verified on Android/MAUI & WPF). 

Please ignore the code comments for now, will update them all once you're happy with this design to minimize churn.

Platforms left to do:
- Maui
  - Windows (just need to re-test)
  - iOS
  - MacOS Catalyst
- WinForms

Eilon commented on src/BlazorWebView/src/Maui/BlazorWebView.cs:
```diff
@@ -23,13 +23,18 @@ public BlazorWebView()
 
 		/// <inheritdoc/>
-		public Func<ExternalLinkNavigationInfo, ExternalLinkNavigationPolicy>? OnExternalNavigationStarting { get; set; } 
+		public event EventHandler<ExternalLinkNavigationEventArgs>? OnExternalNavigationStarting;
```
The event name shouldn't have `On` in it. The method that "invokes" it (the one below) should be `On[EventName]`.

Please confirm that this works in XAML when the BlazorWebView is defined there. The Control.Sample has a commented-out XAML snippet that you should be able to use, or of course start with a new project template (`dotnet new maui-blazor -o MyEventTestApp`).

TanayParikh replied:
> Please confirm that this works in XAML when the BlazorWebView is defined there. The Control.Sample has a commented-out XAML snippet that you should be able to use, or of course start with a new project template (`dotnet new maui-blazor -o MyEventTestApp`).

This is how I tested the WPF XAML bits:
![image](https://user-images.githubusercontent.com/14852843/155403562-45e70970-054e-4cb4-a33c-d27d3629a92b.png)

Policy Configuration:
![image](https://user-images.githubusercontent.com/14852843/155403621-f1861bda-0097-450f-9bff-f4a80dc8b950.png)

Is that what you were looking for?

Eilon replied:
Yes I think that looks right for the XAML.

But the event name still need to change and remove the `On` prefix. The event names are just things like "Pressed", "Loading" etc. and don't start with `On`.

Eilon commented on src/BlazorWebView/src/SharedSource/WebView2WebViewManager.cs:
```diff
@@ -163,18 +176,41 @@ protected virtual void QueueBlazorStart()
+		private void LaunchUri(Uri uri)
+		{
+#if WEBVIEW2_WINFORMS || WEBVIEW2_WPF
+			Process.Start(@"C:\Program Files\Internet Explorer\iexplore.exe", uri.ToString());
```
Definitely can't hard-code this. It's actually even easier: Just call `Process.Start()` with the URL! Windows will Do The Right Thing:tm:

More info: https://docs.microsoft.com/troubleshoot/developer/visualstudio/csharp/general/start-internet-browser#use-the-process-class-start-method-to-start-the-browser

TanayParikh replied:
Hmm I'd tried that but unfortunately it didn't do the right thing 😞 .

```
System.ComponentModel.Win32Exception
  HResult=0x80004005
  Message=An error occurred trying to start process 'https://microsoft.com/' with working directory 'C:\Users\taparik\dev\maui\src\BlazorWebView\samples\BlazorWpfApp\bin\Debug\net6.0-windows'. The system cannot find the file specified.
  Source=System.Diagnostics.Process
  StackTrace:
   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start(String fileName)
   at Microsoft.AspNetCore.Components.WebView.WebView2.WebView2WebViewManager.LaunchUri(Uri uri) in C:\Users\taparik\dev\maui\src\BlazorWebView\src\SharedSource\WebView2WebViewManager.cs:line 209
   at Microsoft.AspNetCore.Components.WebView.WebView2.WebView2WebViewManager.CoreWebView2_NewWindowRequested(Object sender, CoreWebView2NewWindowRequestedEventArgs args) in C:\Users\taparik\dev\maui\src\BlazorWebView\src\SharedSource\WebView2WebViewManager.cs:line 201
   at Microsoft.Web.WebView2.Core.CoreWebView2.OnNewWindowRequested(CoreWebView2NewWindowRequestedEventArgs args)
   at Microsoft.Web.WebView2.Core.CoreWebView2NewWindowRequestedEventHandler.Invoke(ICoreWebView2 source, ICoreWebView2NewWindowRequestedEventArgs args)
```

this was related to https://stackoverflow.com/a/2369171 and hence I'd hardcoded the iexplore path. Would you be open to `Process.Start` with a try/catch for a `Win32Exception` (as seen [here](https://docs.microsoft.com/en-us/troubleshoot/developer/visualstudio/csharp/general/start-internet-browser#complete-code-sample)). If we have `ErrorCode==-2147467259` we use `iexplorer`?

There's nothing special about my setup and I've tried both edge and chrome as the default `http`/`https` apps.

TanayParikh replied:
I believe this may be related to https://github.com/dotnet/runtime/issues/15842.

TanayParikh replied:
Okay yes, was related to `UseShellExecute`. Doing the following resolves the issue:

```csharp
using (var launchBrowser = new Process())
{
    launchBrowser.StartInfo.UseShellExecute = true;
    launchBrowser.StartInfo.FileName = uri.ToString();
    launchBrowser.Start();
}
```

Eilon replied:
Hard-coding iexplore.exe is _never_ the right answer 😁

Eilon commented on src/BlazorWebView/src/Wpf/BlazorWebView.cs:
```diff
@@ -48,9 +48,17 @@ public class BlazorWebView : Control, IAsyncDisposable
 			typeMetadata: new PropertyMetadata(OnServicesPropertyChanged));
+
+		/// <summary>
+		/// The backing store for the <see cref="Services"/> property.
```
Wrong comment.

Eilon reviewed (changes requested):
OK, I think this is looking pretty good (just some fairly small comments). This really hinges on whether this actually works, or not, though 😁 I have no reason to think it wouldn't, but, boy oh boy, this PR gets to deal with all _six_ platforms we support, each with its own unique webview interop, event patterns, launcher APIs, etc.

TanayParikh committed:
Merge branch 'main' into taparik/androidAnchorTagConfigurability

TanayParikh committed:
iOS & Mac Catalyst

TanayParikh committed:
Fix WPF/Winforms Browser Start

TanayParikh committed:
Winforms ExternalNavigationStarting

Eilon commented on src/BlazorWebView/src/Maui/BlazorWebViewHandler.cs:
```diff
@@ -1,15 +1,18 @@
 	public partial class BlazorWebViewHandler
 	{
-		public static PropertyMapper<IBlazorWebView, BlazorWebViewHandler> BlazorWebViewMapper = new(ViewHandler.ViewMapper)
+		private static readonly PropertyMapper<IBlazorWebView, BlazorWebViewHandler> BlazorWebViewMapper = new(ViewMapper)
```
Was this change to private deliberate? The mappers are supposed to be public so that derived types can re-use them. Examples: https://github.com/dotnet/maui/blob/main/src/Core/src/Handlers/Button/ButtonHandler.cs#L19-L41

Eilon replied:
I think making it readonly is good, though interestingly in the one I linked they are not.

Eilon replied:
I logged https://github.com/dotnet/maui/issues/4870 to see if they should be `readonly` across all of MAUI.

TanayParikh replied:
updated to `public static readonly`.

Eilon commented on src/BlazorWebView/src/SharedSource/ExternalLinkNavigationEventArgs.cs:
```diff
@@ -0,0 +1,30 @@
+		/// <summary>
+		/// External <see cref="Uri">URI</see> to be navigated to.
+		/// </summary>
+		public Uri Uri { get; set; }
```
Should this have a setter? I think not?

TanayParikh replied:
I don't see any harm in having it? Wanted to support the hypothetical use case of redirecting the navigation request based on the existing Uri.

Eilon replied:
Doesn't seem necessary because that isn't the point of this. If someone wants a link to go somewhere else, they should just do that to begin with. Also, the current code doesn't even always work that way with a modified event args URL because in at least some cases it uses the original URL and not the one from the event args that may have been modified:

https://github.com/dotnet/maui/pull/4645/files#diff-75c4562a2bc68615b8d28c071b33a112272d1e08e4ba087c461736787ff8f534R199

Eilon commented on src/BlazorWebView/src/SharedSource/ExternalLinkNavigationPolicy.cs:
```diff
@@ -0,0 +1,26 @@
+		OpenInExternalBrowser,
+
+		/// <summary>
+		/// Opens anchor tags <![CDATA[<a>]]> in the WebView. This is not recommended.
```
This still needs some updated comment. Something like:

```suggestion
		/// Opens anchor tags <![CDATA[<a>]]> in the WebView. This is not recommended unless the content of the URL is fully trusted.
```

@blowdart - any suggestion on how to word this? This is related to the threat model I updated earlier. (We'll later have regular docs too, but I think it's nice to have a reasonable doc comment warning here too.)

Eilon reviewed (changes requested):
A few more comments.

TanayParikh committed:
PR Feedback

TanayParikh commented on src/BlazorWebView/src/Maui/BlazorWebViewHandler.cs:
```diff
@@ -1,13 +1,18 @@
-		public static PropertyMapper<IBlazorWebView, BlazorWebViewHandler> BlazorWebViewMapper = new(ViewHandler.ViewMapper)
+		public static readonly PropertyMapper<IBlazorWebView, BlazorWebViewHandler> BlazorWebViewMapper = new(ViewMapper)
 		{
+			// Note the order of the initialization matters. We must set the ExternalNavigationStarting action before starting the core webview.
+			[nameof(IBlazorWebView.ExternalNavigationStarting)] = MapNotifyExternalNavigationStarting,
```
This'll likely be problematic when using multiple property mappers:

https://github.com/dotnet/maui/blob/main/src/Core/src/Handlers/Button/ButtonHandler.cs#L19-L41

looking into how we can make this more generalized.

TanayParikh committed:
Remove ordering dependency during property mapping

TanayParikh commented:
Thanks for the review @Eilon. I believe all the feedback should now be addressed, and all 6 platforms _should_ be working. I'll do a quick validation round across all the platforms before final merge.

TanayParikh commented on src/BlazorWebView/src/SharedSource/ExternalLinkNavigationEventArgs.cs:
```diff
@@ -0,0 +1,30 @@
+	/// Used to provide information about a link (<![CDATA[<a>]]>) clicked within a Blazor WebView.
+	/// 
+	/// Anchor tags with target="_blank" will always open in the default
+	/// browser and the ExternalNavigationStarting event won't be called.
```
Can't cref the ExternalNavigationStarting event because this class is in the `SharedSource` and is used with `ExternalNavigationStarting` in `Maui`, `WinForms` & `WPF`.

TanayParikh commented:
We'll need to add some docs for this new functionality (cc/ @guardrex) alongside maybe having this part of the Preview 14 announcement given we changed defaults(?).

## Instructions

Register to the `ExternalNavigationStarting` event and set the `ExternalLinkNavigationEventArgs.ExternalLinkNavigationPolicy` property to change link handling behavior. The `ExternalLinkNavigationPolicy` enum allows setting link handling behavior to `OpenInExternalBrowser`, `OpenInWebView` and `CancelNavigation`. You may utilize the `ExternalLinkNavigationEventArgs.Uri` property to dynamically set link handling behavior.

Please note, external links are opened in the device default browser, by default. Opening external links within the Blazor WebView is not recommended unless the content is fully trusted.

### MAUI

Add the event handler to the constructor of the `Page` where you construct the `BlazorWebView`:

```csharp
blazorWebView.ExternalNavigationStarting += (sender, externalLinkNavigationEventArgs) =>
{
	externalLinkNavigationEventArgs.ExternalLinkNavigationPolicy = ExternalLinkNavigationPolicy.OpenInWebView;
};
```

### WPF

Add the `ExternalNavigationStarting="Handle_ExternalNavigationStarting"` attribute to the `BlazorWebView` control in your `.xaml` file:

```xaml
<blazor:BlazorWebView HostPage="wwwroot\index.html" Services="{StaticResource services}" x:Name="blazorWebView" ExternalNavigationStarting="Handle_ExternalNavigationStarting" >
```

Add the event handler in your `.xaml.cs` file:

```csharp
private void Handle_ExternalNavigationStarting(object sender, ExternalLinkNavigationEventArgs externalLinkNavigationEventArgs)
{
	externalLinkNavigationEventArgs.ExternalLinkNavigationPolicy = ExternalLinkNavigationPolicy.OpenInWebView;
}
```

### Winforms

In the contructor of the Form containing the `BlazorWebView` control, add the following event registration:

```csharp
blazorWebView.ExternalNavigationStarting += (sender, externalLinkNavigationEventArgs) =>
{
        externalLinkNavigationEventArgs.ExternalLinkNavigationPolicy = ExternalLinkNavigationPolicy.OpenInWebView;
};
```

blowdart commented:
_Opens anchor tags <![CDATA[<a>]]> in the WebView. This is not recommended unless the content of the URL is fully trusted._

Would this apply to all hrefs in the webview? That seems horrible, because local references would be trusted somewhat. If it was external links it'd be easier to describe. "Allows navigation to external links in the WebView. When true this setting can introduce security concerns and should not be enabled unless you can ensure all external links are fully trusted."

Also we now have a habit of naming properties like this as InsecureSomethingOrOther to make it even clearer. OpenInExternalBrowser = false is secure, OpenInExternalBrowser = true potentially isn't. Making it an enum rather than a bool with the option "AllowInsecureOpeningOfLinksInsideWebView" would be clearer.

TanayParikh committed:
@blowdart feedback

TanayParikh commented:
> Would this apply to all hrefs in the webview? That seems horrible, because local references would be trusted somewhat. If it was external links it'd be easier to describe.

Yes, this configurability is specifically designed for **external** URLs, defined as URLs with a `Uri.Host != "0.0.0.0"`.

I've updated the enum to `InsecureOpenInWebView` and also updated the doc comments to better reflect that this is for external links: d4fb97bc4

Eilon commented on src/BlazorWebView/src/SharedSource/ExternalLinkNavigationPolicy.cs:
```diff
@@ -0,0 +1,28 @@
+
+		/// <summary>
+		/// Allows navigation to external links within the Blazor WebView.
+		/// This navigation policy can introduce security concerns and should not be enabled unless you can ensure all external links are fully trusted
```
End sentence with period:

```suggestion
		/// This navigation policy can introduce security concerns and should not be enabled unless you can ensure all external links are fully trusted.
```

TanayParikh committed:
@Eilon feedback

Eilon reviewed (approved):
I approve this message.
